Feat/gsd with queue

b-butler opened the pull request:
## Description

Add a `hoomd.write.Deque` class which stores a `std::deque` of logger and trajectory data up to a specified maximum size `N`. The deque pushes data to the front and pops old data off the back of the deque. The `analyze` method only collects and removes data as necessary. All writes are triggered by the `dump` method exposed in Python. The purpose of this class is to allow selective high frequency data storing, by allowing users to determine when to write data.

One foreseen pattern of use would be to create a custom action which composes a `Deque` object calling `dump` when the specified conditions have been met.

## Motivation and context

To provide a means for high-performance, high-frequency writes in HOOMD.

## How has this been tested?

Tests based on the GSD tests were added.

## Change log

<!-- Propose a change log entry. -->
```
Added: `hoomd.write.Deque` class for enabling selective high-frequency frame writing.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
refactor: Separate gsd frame writing from analyze.

This will more easily allow for derived classes of GSDDumpWriter since
they can choose when to write or not.

b-butler committed:
feat: Add C++ GSDDequeWriter

This class provided a triggerable multi frame write.

joaander committed:
Embed latest gsd implementation.

joaander committed:
Add GSD.flush

joaander committed:
Add necessary flush() calls to gsd unit tests.

joaander committed:
Add maximum_write_buffer_size to GSD.

Also move initFileIO to the constructor. This prevents the need for
double caching of the buffer size prior to opening the file.

joaander committed:
Flush on calls to GSD.write.

Ensure that we do not rely on Python's garbage collector to ensure
that the file is written after the call to write() ends.

joaander committed:
Run pre-commit.

joaander committed:
Scale index buffer size with write buffer size.

Also check for errors in new GSD API calls.

joaander committed:
Install SIGTERM handler.

joaander committed:
Sync gsd.c with gsd package.

b-butler committed:
refactor: Make GSDDumpWriter more amenable to subclassing

- Move initFileIO to protected
- Add method for querying file initialization

b-butler committed:
feat: Add Deque writer to Python.

joaander committed:
Update migration guide.

b-butler committed:
test: Test existing Deque writer functionality.

b-butler committed:
refactor: Move writeLogQuantities to C++

b-butler committed:
feat: Add logging ability to Deque.

b-butler committed:
refactor: Remove now redundant IO check in Deque

b-butler committed:
refactor: Make m_handle available to subclasses (GSDDumpWriter).

b-butler committed:
fix: Write first frame in GSDDequeWriter

While this is unexpected, it is required to keep performance high as
until the first write for a file occurs, all data is recorded and sorted
in a GSDDumpWriter::GSDFrame object. By writing the first frame, we
prevent this. However, we then don't know if the first frame should be
included in the actual analysis.

b-butler committed:
test: Update Deque tests.

All tests including logging tests now pass.

b-butler committed:
fix: Deque.__init__ correctly call base class constructor.

b-butler committed:
refactor: Rename Deque writer to Burst

b-butler committed:
doc: Add Burst to documentation.

b-butler committed:
Merge branch 'trunk-major' into feat/gsd-with-queue

AlainKadar commented on hoomd/md/pytest/test_burst_writer.py:
```diff
@@ -0,0 +1,202 @@
+
+
+@pytest.fixture(scope='function')
+def hoomd_snapshot(lattice_snapshot_factory):
```
Do we want to call this `frame` instead of `snapshot`? Or better to keep consistent with the name of the module? Are there plans to rename this module to `frame` in the future?

joaander replied:
No, these are `hoomd.Snapshot` objects. I renamed `gsd.Snapshot` to `gsd.Frame` to avoid confusion with `hoomd.Snapshot`.

AlainKadar reviewed (approved):
I just had one suggestion

b-butler commented:
@joaander The test errored due to timeout which is currently set at 3000 second or 50 minutes but MPI tests finish in 30 seconds on hodges. Also, MPI tests for the neighbor list are erroring for so reason on hodges. Will look into that.

joaander commented:
> @joaander The test errored due to timeout which is currently set at 3000 second or 50 minutes but MPI tests finish in 30 seconds on hodges. Also, MPI tests for the neighbor list are erroring for so reason on hodges. Will look into that.

I can't look at it right now. Maybe later this week. If you can find a configuration that reproduces the deadlock, I usually debug them by 1) Running the tests with mpirun and wait for the deadlock. 2) Attach a debugger to the processes and see where each rank is waiting.

b-butler committed:
fix: Correctly set m_write_initial_frame with MPI

b-butler committed:
test: Fix the burst writer tests.

b-butler commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

b-butler committed:
fix: Non-MPI GSDDequeWriter m_write_initial_frame setting

b-butler committed:
test: Reduce runtime of Burst writer test

b-butler commented:
The deadlocking is solved. I forgot to not call gsd functions on all ranks.

joaander commented on hoomd/GSDDumpWriter.cc:
```diff
@@ -342,56 +352,57 @@ void GSDDumpWriter::analyze(uint64_t timestep)
         }
 
     // emit on all ranks, the slot needs to handle the mpi logic.
     m_write_signal.emit(m_handle);
```
If this code was used, it would be a problem for the burst writer. However, this has now been replaced by `Logger`. I opened #1554 to remove these dead code paths. No need to make any changes in this pull request.

joaander replied:
By `this`, I refer to `m_write_signal.emit`.

joaander commented on hoomd/write/gsd_burst.py:
```diff
@@ -0,0 +1,86 @@
+class Burst(GSD):
+    r"""Write last :math:`N` stored frames at user trigger in the GSD format.
+
+    This class stores the last :math:`N` frames in an interal deque
```
```suggestion
    This class stores up to the last :math:`N` frames in an interal deque
```

joaander commented on hoomd/write/gsd_burst.py:
```diff
@@ -0,0 +1,86 @@
+            to `None`.
+        max_burst_frames (int): The maximum number of frames to store before
+            between writes. -1 represents no limit. Defaults to -1.
+
```
Add a note that when there are no frames in the GSD file, `Burst` always one frame the first time it is triggered to capture all the static quantities.

joaander commented on hoomd/write/gsd_burst.py:
```diff
@@ -0,0 +1,86 @@
+    This class stores the last :math:`N` frames in an interal deque
+    which only writes the frames out when `dump` is called. When the writer is
+    triggered and the next frame would result in :math:`N + 1` frames being
+    stored, the oldest frame is popped off the deque and the new frame is
+    pushed to the front of the deque.
```
```suggestion
    stored, the oldest frame is removed from the deque and the new frame is
    added.
```
Also, consider replacing `deque` (an implementation detail) with a more generic term such as `queue` or `buffer` to help users understand.

joaander reviewed (changes requested):
Thanks! The implementation is very clean. I have a few suggestions to improve the documentation.

b-butler committed:
doc: Improve Burst writer documentation

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
doc: Add warning about initial Burst write.

This behavior is not intuitive but required for good performance.

b-butler commented:
@joaander I just thought of a potential mitigation of the initial write from the burst writer. We could either
* provide a method to write similar to `GSD.write`
* or show how to get this to run at the very beginning of a simulation with an `And` trigger.

What do you think?

joaander commented:
Yes, we could encourage users to use `GSD.write` to write the initial file and then open it in append mode. We could also enforce this behavior by making it an error for `BurstWriter` to create files or append to files with 0 frames.

b-butler committed:
refactor: Error by default when using empty file for Burst.

Adds option to run immediately upon attaching for those still wishing to
use this.

b-butler committed:
test: Update Burst tests

b-butler committed:
test: Test new write_on_start Burst attribute

b-butler committed:
refactor: Move writers to end of Operations._schedule

This improve the ability for writers to immediately upon attaching write
log data depending on one or more computes or tuners.

joaander committed:
Revise docs.

joaander commented on hoomd/write/gsd_burst.py:
```diff
@@ -0,0 +1,106 @@
+            to `None`.
+        max_burst_frames (int): The maximum number of frames to store before
+            between writes. -1 represents no limit. Defaults to -1.
+        write_at_start (bool): When ``True`` **and** the file does not exist or
```
`write_at_start` seems like the wrong name for this. Consider using `write_initial_frame`.

b-butler replied:
Decided offline that the current name is okay.

joaander reviewed (commented):
I pushed some documentation revisions directly to the branch. Please review and make any further edits as needed. It looks like you need to merge in conflicts from `trunk-major`.

I have one comment on parameter naming to think about. Other than that, this is good to merge.

b-butler committed:
Merge branch 'trunk-major' into feat/gsd-with-queue

b-butler committed:
fix: mistake in previous merge

joaander reviewed (approved):
Thanks!
